src/sleep.rs

```diff
@@ -0,0 +1,763 @@
+pub trait RtcWakeupPins {
+    type WakeupPin: RtcWakeupPinTrait;
+    type Iterator<'a>: Iterator<Item = i32>
+    where
+        Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_>;
+}
+#[cfg(not(any(esp32, esp32s2, esp32s3)))]
+pub trait RtcWakeupPins {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<P> RtcWakeupPins for P
+where
+    P: RtcWakeupPinTrait,
+    P::PinDriver: RtcPinDriver,
+{
+    type WakeupPin = P;
+
+    type Iterator<'a> = core::iter::Once<i32> where Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_> {
+        core::iter::once(self.pin())
+    }
+}
+
+pub struct EmptyRtcWakeupPins<A>(PhantomData<A>);
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<A> EmptyRtcWakeupPins<A> {
+    pub fn chain<O>(other: O) -> ChainedRtcWakeupPins<Self, O>
+    where
+        A: RtcWakeupPinTrait,
+        O: RtcWakeupPins<WakeupPin = A>,
+    {
+        ChainedRtcWakeupPins {
+            first: Self(PhantomData),
+            second: other,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<A> RtcWakeupPins for EmptyRtcWakeupPins<A>
+where
+    A: RtcWakeupPinTrait,
+{
+    type WakeupPin = A;
+
+    type Iterator<'a> = core::iter::Empty<i32> where Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_> {
+        core::iter::empty()
+    }
+}
+
+#[cfg(not(any(esp32, esp32s2, esp32s3)))]
+impl<A> RtcWakeupPins for EmptyRtcWakeupPins<A> {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub struct ChainedRtcWakeupPins<F, S> {
+    first: F,
+    second: S,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<F, S> ChainedRtcWakeupPins<F, S> {
+    pub fn chain<O>(self, other: O) -> ChainedRtcWakeupPins<Self, O>
+    where
+        F: RtcWakeupPins,
+        S: RtcWakeupPins,
+        O: RtcWakeupPins,
+    {
+        ChainedRtcWakeupPins {
+            first: self,
+            second: other,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<F, S> RtcWakeupPins for ChainedRtcWakeupPins<F, S>
+where
+    F: RtcWakeupPins,
+    S: RtcWakeupPins,
+{
+    type WakeupPin = F::WakeupPin;
+
+    type Iterator<'a> = core::iter::Chain<F::Iterator<'a>, S::Iterator<'a>> where Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_> {
+        self.first.iter().chain(self.second.iter())
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub struct RtcWakeupPin<'d, P, M>
+where
+    P: InputPin + 'd,
+    M: RTCMode + 'd,
+{
+    pub pindriver: &'d PinDriver<'d, P, M>,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<'d, P, M> RtcWakeupPinTrait for RtcWakeupPin<'d, P, M>
+where
+    P: InputPin + 'd,
+    M: RTCMode + 'd,
+{
+    type PinDriver = PinDriver<'d, P, M>;
+
+    fn pin(&self) -> i32 {
+        self.pindriver.pin()
+    }
+}
+
+/// Will wake up the CPU based on changes in GPIO pins. Is only available for light sleep.
+/// It takes PinDriver as input, which means that any required configurations on the pin
+/// can be done before adding it to the GpioWakeup struct.
+pub struct GpioWakeup<P>
```

Comment:
All suggestions for simplifications for the RTC subsystem apply here as well, as the approach is the same.